opsimulator/opsimulator.go

```diff
@@ -189,28 +177,26 @@ func handler(proxy *httputil.ReverseProxy) http.HandlerFunc {
 			return
 		}
 
-		body, err := io.ReadAll(r.Body)
-		if err != nil {
-			http.Error(w, "Failed to read request body", http.StatusInternalServerError)
-			return
-		}
-		defer r.Body.Close()
+		// setup an intermediate buffer so the request body is inspectable
+		var buf bytes.Buffer
+		body := io.TeeReader(r.Body, &buf)
+		r.Body = io.NopCloser(&buf)
 
-		var req JSONRPCRequest
-		err = json.Unmarshal(body, &req)
+		// decode the fields we're interested in inspecting
+		msgs, err := readJsonMessages(body)
 		if err != nil {
-			http.Error(w, "Failed to parse JSON-RPC request", http.StatusBadRequest)
+			http.Error(w, fmt.Sprintf("Failed to parse JSON-RPC request: %s", err), http.StatusBadRequest)
 			return
 		}
 
 		// TODO(https://github.com/ethereum-optimism/supersim/issues/55): support batch txs
 
-		if req.Method == "eth_sendRawTransaction" {
-			checkInteropInvariants()
+		for _, msg := range msgs {
+			if msg.Method == "eth_sendRawTransaction" {
+				checkInteropInvariants()
```

Comment:
if one of the messages in the batch contains a tx that does not pass the interop invariant check, should we error out the entire batch or just that tx?

Reply:
we can error just that tx. each individual rpc within the batch has an `Error` field that we can populate

Reply:
does this get tricky since we are also forwarding to a reverse proxy (anvil)? for example, we would need to stop txs that fail the interop checks from being forwarded to the reverse proxy, and then combine the failed txs with the response for the txs that were sent to the reverse proxy.

Reply:
we could filter out the failed ones when forwarding to interop but in the response populate the error if that makes sense